### PR TITLE
default.nix: Fix neat-interpolation override

### LINKED ISSUE #706: CI breakage with "NixOS-unstable channel, GHC 8.6.5" and "Nixpkgs-unstable channel, default GHC (8.8)"
Both fail with `error: attribute 'neat-interpolation_0_5_1_1' missing, at /home/runner/work/hnix/hnix/default.nix:229:55`.

See e.g. https://github.com/haskell-nix/hnix/pull/705/checks.

### PR DESCRIPTION
**Pull request by sjakobi**
Recent versions of Nixpkgs don't provide neat-interpolation_0_5_1_1.

Defining the derivation explicitly should make this solution more
robust against future Nixpkgs changes.

Fixes #706.

### CODE REVIEW

**Comment by sjakobi** [comment]
This PR is a rebased and slightly tweaked version of https://github.com/haskell-nix/hnix/pull/680 BTW.

**Comment by sjakobi** [comment]
The [Nixpkgs, macOS / NixOS-unstable, default GHC (8.8) job](https://github.com/haskell-nix/hnix/pull/707/checks?check_run_id=1087715302) has timed out for some reason. Any ideas why?

I'll retry.

**Comment by sjakobi** [comment]
> The [Nixpkgs, macOS / NixOS-unstable, default GHC (8.8) job](https://github.com/haskell-nix/hnix/pull/707/checks?check_run_id=1087715302) has timed out for some reason. Any ideas why?
> 
> I'll retry.

Looks like that job is going to time out again. @Anton-Latukha do you have an idea how to fix this?

**Comment by sjakobi** [comment]
Retrying again…

**Comment by Anton-Latukha** [comment]
`master` branch builds fine: https://github.com/haskell-nix/hnix/actions?query=workflow%3A%22Nixpkgs%2C+macOS%22, so that means that this branch CI build does the computation but does not terminate.

So it is something between the override and Nix slipping into nonterminating computation.

Nix does type checking in an ad-hock manner, it may slip into nonterminating computation, that is why Nix does not proclaims the termination guarantee, it has some early detection and termination in Nix interpreted, but that is not ideal.

That is what I am talking about.

What is the problem with the line you try to solve? There is no `neat-interpolation_0_5_1_1` in Nixpkgs code, because there is `neat-interpolation_0_5_1_2`, which you try to find.

Why the most explicit hardcode on a version is more robust than doing the https://search.nixos.org/packages?query=neat-interpolation&from=0&size=30&sort=relevance&channel=unstable and changing one character to `neat-interpolation_0_5_1_(1->2)`?

If we leave `neat-interpolation = pkgs.haskell.lib.dontCheck super.neat-interpolation_x_x_x_x` package, it is equivalent to explicit override you try to do. They both are hardcode on a version, with explicit hardcode being more problematic, because to update the sha256 needs to be calculated and supplied, and it is much more code.

Hydra does not prune the cache, so we magically receive the caches old versions of packages that are lo longer explicitly in Nixpkgs (I do not know the process of that, but they work and arrive, most probably Nix ecosystem simply detects that that is a particular old version request and gives it).

People do the support of Haskell Nixpkgs packages inside Nixpkgs for us, so tracking the Nixpgkgs for packages - we get a proper chunk of work offloaded to them, they manage Nixpkgs specifics and package-special overrides for users of the store.

**Comment by sjakobi** [comment]
Thanks for the feedback, @Anton-Latukha! :)

The problem that I'm trying to solve is #706. Apparently some Nixpkgs revisions that we're testing in CI don't provide `neat-interpolation_0_5_1_1` which we're currently relying on.

Now we could just use `neat-interpolation_0_5_1_2`, but that means that CI would very likely break again when yet another `neat-interpolation` version is released.

So I would prefer to handle this override similarly to the `data-fix` and `prettyprinter` overrides below. Why doesn't this work? What's the difference between overriding `data-fix` with an explicit derivation and overriding `neat-interpolation`? 

If you could work this out, I would greatly appreciate that! :)

**Comment by sjakobi** [comment]
> `master` branch builds fine: https://github.com/haskell-nix/hnix/actions?query=workflow%3A%22Nixpkgs%2C+macOS%22, so that means that this branch CI build does the computation but does not terminate.

Why isn't `master` affected by the missing `neat-interpolation_0_5_1_1` issue BTW?

**Comment by sjakobi** [comment]
Could the timeout be a `macOS`-specific problem? How about simply removing that job?

**Comment by Anton-Latukha** [comment]
What I basically do on Nixpkgs problems, I give it time sometimes. Nixpkg default processes differ from `peti` autogeneration path, Haskell-Nixpkgs path is much faster and more consistent in version updates, if `megaparsec` arrived into Nixpkgs, and it demands the `neat-interpolation 0.5.1.2` - it means that `neat-interpolation_0_5_1_2` is present in Nixpkgs.

There always would be a couple of days lag of Hackage packages spreading into other stores, Nixpkgs also. So if you found that package updated - no reason to run and try override Nix configuration, since in couple of days those overrides would be redundant.

Nixpkgs syncs most packages with Hackage every night. Except for the manually hardcoded packages&versions.
Basically Nix configuration should lag behind Hackage. So people arrive into Nixpkgs, manage manual overrides and merge them.
And since our Nix configuration automatically generates from Cabal.
That means that our Cabal configuration should be a couple of days (half a geek mostly, weak mostly guaranteed, up to two weeks timeout) behind Hackage version package updates, so when you get to updating - they are already in Nixpkgs.

**Comment by Anton-Latukha** [comment]
If there is something about Nix package version updates - it needs to be looked, if there is PR in Nixpkgs with `haskell` tag that addresses that, and if no - upstream update into Nixpkgs, that is the most simple and effective active way, the lazy way - is to look at it after a week, the most lazy way - react when the `master` CI drops.

**Comment by sjakobi** [comment]
@Anton-Latukha Do I understand you correctly that we should just wait for a few days until Nixpkgs is somehow "fixed" and our CI starts working again?

Such a dependency on Nixpkgs processes sounds pretty terrible to me. Why don't we simply remove the macOS job for now? The other ones don't seem to have this problem.

**Comment by Anton-Latukha** [comment]
> Could the timeout be a `macOS`-specific problem? How about simply removing that job?

Why it is the only approach of you.

CI repeatedly shows that PR change in that conditions does not terminate. That information is useful. Isn't maintaining Nix environment is the agenda of the work, and CI shows that this change, in this case, does not terminate.

**Comment by Anton-Latukha** [comment]
https://github.com/haskell-nix/hnix/pull/707#issuecomment-690295702

Ok, from that I understood.

It is an internal lag between Nixpkgs revisioins.
CI already requires only `nixos-unstable`, others are optional and for information, optional builds taint the state report (it directs attention to them), but it is not required and CI permits the merges.

**Comment by sjakobi** [comment]
> > Could the timeout be a `macOS`-specific problem? How about simply removing that job?
> 
> Why it is the only approach of you.
> 
> CI repeatedly shows that PR change in that conditions does not terminate. That information is useful.

@Anton-Latukha I don't know how to interpret the CI output. Can you help me understand it?

If you have a better solution for #706 in mind, please do provide a fix.

**Comment by Anton-Latukha** [comment]
But, as we see to merge this PR, the `mac-OS` is required, and it is the same `nixos-unstable`, so CI requires consistency of HNix config and `nixos-unstable` channel to pass the PR into merge.

**Comment by sjakobi** [comment]
> But, as we see to merge this PR, the `mac-OS` is required, and it is the same `nixos-unstable`, so CI requires consistency of HNix config and `nixos-unstable` channel to pass the PR into merge.

What does "consistency" mean here? Can you fix it?

**Comment by Anton-Latukha** [comment]
Nix channels are FIFO process (`master -> nixpkgs-unstable -> nixos-unstable -> nixos-20.09 -> nixos-20.03`), which means different state for every channel.

**Comment by sjakobi** [comment]
> Nix channels are FIFO process (`master -> nixpkgs-unstable -> nixos-unstable -> nixos-20.09 -> nixos-20.03`), which means different state for every channel.

I have no idea what to do with this information nor how it might be related to this PR.

**Comment by Anton-Latukha** [comment]
Opened #710, that definitely should solve the need for this kind of overrides.

**Comment by Anton-Latukha** [comment]
Solved by pinning Nixpkgs.
